velox/functions/sparksql/aggregates/DecimalSumAggregate.h

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include "velox/exec/Aggregate.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::functions::aggregate::sparksql {
+
+struct DecimalSum {
+  int128_t sum{0};
+  int64_t overflow{0};
+  bool isEmpty{true};
+};
+
+template <typename TInputType, typename TResultType>
+class DecimalSumAggregate : public exec::Aggregate {
```

Comment:
Can this function be implemented using simple api? CC: @kagamiori @laithsakka

Reply:
@mbasmanova Can we add an additional parameter to the `SimpleAggregateAdapter`? I need to pass the "decimal sumType" to the `DecimalSumAggregate`.

Reply:
@liujiayi771 I'm not sure I understand this question. Would you elaborate a bit?

Reply:
@liujiayi771 This PR seems to have a lot of complex code copy-pasted. It is hard to review. It would be great if it could be simplified using Simple API.

CC: @kagamiori

Reply:
Hi @mbasmanova and @liujiayi771, Unfortunately, this function currently cannot be implemented through the simple interface. The reason is that the aggregation function's extractAccumulator and extractValues methods need to use the `sumType_` that comes from `resultType` of the function (`sumType_` provides the precision and scale of the result Decimal). However, the SimpleAggregateAdapter doesn't expose `resultType` to the UDAF author. 

It looks like requiring to know the result type inside functions' logic is a common pattern among Decimal functions. This implies that we would need to expose `resultType` to the UDAF authors in the simple UDAF interface.

Reply:
Hi @kagamiori, `sumType_` does not only come from `resultType`, but also from `inputType` when the input is intermediate data. I add a getDecimalSumType method [here](https://github.com/facebookincubator/velox/pull/5372/files#diff-cc268331a2906e53fb6519e62ebaa63353fb61412674059b34dff2dff0fe06feR29).

@mbasmanova If this agg cannot use `SimpleAggregateAdapter`, can the current implementation be retained, or is there any other method to simplify a lot of repetitive code? When developing, I mainly referred to the `DecimalAggregate` of prestosql.

Reply:
@liujiayi771 It seems to me that sumType can be derived from result type. It is always DECIMAL(r_precision, r_scale), no? 

if resultType->isRow(), then resultType->childAt(0)
otherwise resultType 

> sumType_ does not only come from resultType, but also from inputType when the input is intermediate data. I add a getDecimalSumType method here.

For some reason, I do not see getDecimalSumType using inputType. It seems to use only result type.

TypePtr getDecimalSumType(
    const TypePtr& resultType,
    core::AggregationNode::Step step) {

Reply:
@mbasmanova Thank you for your patience, I will try to use `SimpleAggregateAdapter` to implement this agg function.